File: leetcode_java/713.SubarrayProductLessThanK.java
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1) return 0;
        int product = 1, count = 0;

        for (int i = 0, left = 0; i < nums.length; ++ i) {
            product *= nums[i];
            while (product >= k) {
                product /= nums[left];
                ++ left;
            }
            count += i - left + 1;
        }
        return count;
    }
}

/**
713. Subarray Product Less Than K

Given an array of integers nums and an integer k, return the number of contiguous subarrays 
where the product of all the elements in the subarray is strictly less than k.
 */

File: leetcode_javascript/713.SubarrayProductLessThanK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numSubarrayProductLessThanK = function(nums, k) {
    if (k <= 1) return 0;
    const n = nums.length;
    let count = 0, product = 1;

    for (let i = 0, begin = 0; i < n; ++ i) {
        product *= nums[i];

        while (product >= k) {
            product /= nums[begin];
            ++ begin;
        }
        count += i - begin + 1;
    }
    return count;
};
